// @dmalbuquerque/cpf-cnpj-validator/dist/cnpj-bkp.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const CNPJ_LENGTH = 14;
const ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
/**
 * Gera um número de CNPJ.
 *
 * Esta função gera um CNPJ válido. O CNPJ gerado pode ser retornado com diferentes tipos: numérico ou alfanumérico (letras e números).
 * Além disso, pode ser retornado com ou sem formatação.
 *
 * @param {CNPJOptions} [options] - Opções para a geração do CNPJ.
 * @param {('numeric' | 'alfanumeric')} [options.type] - O tipo de CNPJ a ser gerado.
 * Pode ser:
 * - `'numeric'`: Gera um CNPJ clássico, composto apenas por números.
 * - `'alfanumeric'`: Gera um CNPJ com caracteres alfanuméricos.
 * Se não for especificado, o padrão será `'numeric'`.
 * @param {boolean} [options.formatted] - Se verdadeiro, retorna o CNPJ formatado. Caso contrário, retorna o CNPJ sem formatação.
 *
 * @returns {string} O CNPJ gerado, podendo ser formatado ou não, conforme as opções fornecidas.
 *
 * @example
 * // Retorna um CNPJ clássico numérico sem formatação, como '25143815000150'
 * generate({ type: 'numeric' });
 *
 * @example
 * // Retorna um CNPJ clássico numérico formatado, como '25.143.815/0001-50'
 * generate({ type: 'numeric', formatted: true });
 *
 * @example
 * // Retorna um CNPJ alfanumérico sem formatação, como 'OGZP0N77444Y42'
 * generate({ type: 'alfanumeric' });
 *
 * @example
 * // Retorna um CNPJ alfanumérico formatado, como 'OG.ZP0.N77/444Y-42'
 * generate({ type: 'alfanumeric', formatted: true });
 */
function generate(options) {
    let cnpj = undefined;
    if (options?.type === "alfanumeric") {
        cnpj = generateAlphanumeric();
    }
    else {
        cnpj = generateNumeric();
    }
    return options?.formatted ? cnpj = format(cnpj) : cnpj;
}
/**
 * Gera um CNPJ numérico válido (clássico).
 *
 * Esta função gera um CNPJ numérico válido, composto apenas por números.
 * O número gerado é composto por uma base numérica gerada e os dois dígitos verificadores calculados.
 *
 * @returns {string} O CNPJ numérico gerado, como uma string de 14 dígitos, sem formatação (apenas números).
 *
 * @example
 * // Retorna um CNPJ numérico válido, como '25143815000150'
 * generateNumeric();
 */
function generateNumeric() {
    const baseCNPJ = generateBaseNumeric();
    const [firstDigit, secondDigit] = calculateCheckDigits(baseCNPJ);
    return [...baseCNPJ, firstDigit, secondDigit].join("");
}
/**
 * Gera um CNPJ alfanumérico válido.
 *
 * Esta função gera um CNPJ alfanumérico válido. O CNPJ gerado é composto por uma base alfanumérica
 * e os dois dígitos verificadores calculados com base nela.
 *
 * @returns {string} O CNPJ alfanumérico gerado, como uma string de 14 caracteres, contendo letras e números, sem formatação.
 *
 * @example
 * // Retorna um CNPJ alfanumérico válido, como 'OGZP0N77444Y42'
 * generateAlphanumeric();
 */
function generateAlphanumeric() {
    const baseCNPJ = generateBaseAlphanumeric();
    const [firstDigit, secondDigit] = calculateCheckDigits(baseCNPJ);
    const alphanumericCNPJ = [...baseCNPJ, firstDigit, secondDigit];
    return alphanumericCNPJ.join("");
}
/**
 * Gera a base numérica de um CNPJ.
 *
 * Esta função gera uma base numérica aleatória composta por 12 dígitos (números de 0 a 9),
 * que são utilizados como a primeira parte do CNPJ, antes dos dois dígitos verificadores.
 *
 * @returns {number[]} Um array de 12 números (de 0 a 9), representando a base numérica do CNPJ.
 *
 * @example
 * // Retorna algo como [2, 4, 6, 1, 5, 3, 9, 7, 8, 0, 1, 4]
 * generateBaseNumeric();
 */
function generateBaseNumeric() {
    return Array.from({ length: 12 }, () => Math.floor(Math.random() * 10));
}
/**
 * Gera a base alfanumérica de um CNPJ.
 *
 * Esta função gera uma base alfanumérica aleatória composta por 12 caracteres,
 * sendo que cada posição pode conter um número (de 0 a 9) ou uma letra (A-Z).
 * A escolha entre número e letra é feita aleatoriamente para cada posição.
 *
 * @returns {(number | string)[]} Um array de 12 elementos, que podem ser números ou letras,
 * representando a base alfanumérica do CNPJ.
 *
 * @example
 * // Retorna algo como [2, 'F', 5, 'G', 8, 'B', 0, 'A', 1, 'C', 3, 'D']
 * generateBaseAlphanumeric();
 */
function generateBaseAlphanumeric() {
    const base = [];
    for (let i = 0; i < 12; i++) {
        const randomChoice = Math.random() < 0.5;
        if (randomChoice) {
            base.push(Math.floor(Math.random() * 10));
        }
        else {
            const letter = ALPHABET.charAt(Math.floor(Math.random() * ALPHABET.length));
            base.push(letter);
        }
    }
    return base;
}
/**
 * Calcula os dois dígitos verificadores de um CNPJ.
 *
 * A função utiliza a base do CNPJ e calcula os dois dígitos verificadores necessários
 * para completar o número do CNPJ. O primeiro dígito é calculado com a base original,
 * e o segundo dígito é calculado após a adição do primeiro dígito.
 *
 * @param baseCNPJ - A base do CNPJ, composta por números ou letras, sem os dígitos verificadores.
 *
 * @returns {number[]} Um array com os dois dígitos verificadores do CNPJ.
 *
 * @example
 * // Retorna algo como [5, 8]
 * calculateCheckDigits([2, 'F', 5, 'G', 8, 'B', 0, 'A', 1, 'C', 3, 'D']);
 */
function calculateCheckDigits(baseCNPJ) {
    const firstDigit = calculateCheckDigit(baseCNPJ);
    const secondDigit = calculateCheckDigit([...baseCNPJ, firstDigit]);
    return [firstDigit, secondDigit];
}
/**
 * Calcula o dígito verificador de um CNPJ ou CPF a partir dos números ou letras fornecidos.
 *
 * A função utiliza a soma ponderada dos números, considerando o peso de cada posição,
 * e aplica a fórmula do módulo 11 para calcular o dígito verificador.
 * Caso o resto da divisão seja menor que 2, o dígito verificador será 0, caso contrário,
 * o dígito será 11 menos o resto.
 *
 * @param digits - Um array de números ou letras que representam a base do CNPJ ou CPF.
 *
 * @returns {number} O dígito verificador calculado.
 *
 * @example
 * // Retorna 5
 * calculateCheckDigit([2, 5, 1, 4, 3, 8, 1, 5, 0, 0, 1, 5]);
 */
function calculateCheckDigit(digits) {
    let sum = 0;
    let weight = 2;
    for (let i = digits.length - 1; i >= 0; i--) {
        const digit = digits[i];
        const numericDigit = typeof digit === "string" ? convertLetterToNumber(digit) : digit;
        sum += numericDigit * weight;
        weight = weight === 9 ? 2 : weight + 1;
    }
    const rest = sum % 11;
    return rest < 2 ? 0 : 11 - rest;
}
/**
 * Formata um número de CNPJ para o formato padrão com pontuação.
 *
 * A função recebe um CNPJ como string sem formatação e o converte para o formato padrão:
 * XX.XXX.XXX/XXXX-XX.
 *
 * @param cnpj - O CNPJ a ser formatado, fornecido como uma string sem pontuação.
 *
 * @returns {string} O CNPJ formatado.
 *
 * @example
 * // Retorna "25.143.815/0001-50"
 * format("25143815000150");
 */
function format(cnpj) {
    return cnpj.replace(/(.{2})(.{3})(.{3})(.{4})(.{2})/, "$1.$2.$3/$4-$5");
}
/**
 * Converte uma letra para um número correspondente conforme a tabela do CNPJ alfanumérico.
 *
 * A função mapeia letras de A a Z para números de 17 a 42, conforme o padrão utilizado para
 * a geração de CNPJ alfanumérico.
 *
 * @param digit - A letra a ser convertida para número.
 *
 * @returns {number} O número correspondente à letra, ou 0 caso a letra não seja válida.
 *
 * @example
 * // Retorna 17
 * convertLetterToNumber("A");
 *
 * // Retorna 42
 * convertLetterToNumber("Z");
 */
function convertLetterToNumber(digit) {
    const letterMap = {
        A: 17,
        B: 18,
        C: 19,
        D: 20,
        E: 21,
        F: 22,
        G: 23,
        H: 24,
        I: 25,
        J: 26,
        K: 27,
        L: 28,
        M: 29,
        N: 30,
        O: 31,
        P: 32,
        Q: 33,
        R: 34,
        S: 35,
        T: 36,
        U: 37,
        V: 38,
        W: 39,
        X: 40,
        Y: 41,
        Z: 42,
    };
    return letterMap[digit] || 0;
}
/**
 * Valida um CNPJ clássico (numérico) verificando seu formato e seus dígitos verificadores.
 *
 * A função limpa a string do CNPJ, valida o formato e calcula os dígitos verificadores
 * para garantir que o CNPJ fornecido seja válido.
 *
 * @param cnpj - O CNPJ a ser validado, que pode estar no formato com ou sem pontuação.
 *
 * @returns {boolean} Retorna `true` se o CNPJ for válido e `false` caso contrário.
 *
 * @example
 * // Retorna true se o CNPJ for válido
 * validateNumeric("25143815000150");
 *
 * // Retorna false se o CNPJ for inválido
 * validateNumeric("25143815000140");
 */
function validateNumeric(cnpj) {
    const cleanedCNPJ = cleanCNPJ(cnpj, true);
    if (!isValidFormat(cleanedCNPJ))
        return false;
    const digits = cleanedCNPJ.split("").map(Number);
    const baseCNPJ = digits.slice(0, 12);
    const [firstDigit, secondDigit] = calculateCheckDigits(baseCNPJ);
    return firstDigit === digits[12] && secondDigit === digits[13];
}
/**
 * Valida um CNPJ alfanumérico através dos dígitos verificadores.
 *
 * A função converte o CNPJ alfanumérico em uma sequência de números, calcula os dígitos verificadores
 * e os compara com os dois últimos dígitos do CNPJ fornecido para garantir sua validade.
 *
 * @param cnpj - O CNPJ alfanumérico a ser validado, que pode estar no formato com ou sem pontuação.
 *
 * @returns {boolean} Retorna `true` se o CNPJ alfanumérico for válido e `false` caso contrário.
 *
 * @example
 * // Retorna true se o CNPJ alfanumérico for válido
 * validateAlfanumeric("OGZP0N77444Y42");
 *
 * // Retorna false se o CNPJ alfanumérico for inválido
 * validateAlfanumeric("OGZP0N77444Y40");
 */
function validateAlfanumeric(cnpj) {
    const digits = convertCnpj(cnpj.length === 14 ? cnpj.slice(0, 12) : cnpj);
    const [dv1, dv2] = calculateCheckDigits(digits);
    const twoFinalDigitsOriginal = parseInt(cnpj.slice(-2));
    const twoFinalDigitsCalculated = parseInt(`${dv1}${dv2}`);
    return twoFinalDigitsOriginal === twoFinalDigitsCalculated;
}
/**
 * Valida um CNPJ (numérico ou alfanumérico) verificando seu formato e seus dígitos verificadores.
 *
 * A função remove qualquer caractere não alfanumérico do CNPJ e, em seguida, verifica se o CNPJ tem o
 * comprimento correto. Dependendo do tipo de CNPJ (numérico ou alfanumérico), a função valida o CNPJ
 * chamando a função apropriada para cada tipo.
 *
 * @param cnpj - O CNPJ a ser validado, que pode estar no formato com ou sem pontuação.
 *
 * @returns {boolean} Retorna `true` se o CNPJ for válido (numérico ou alfanumérico) e `false` caso contrário.
 *
 * @example
 * // Retorna true se o CNPJ for válido
 * isValid("25.143.815/0001-50");
 *
 * // Retorna false se o CNPJ for inválido
 * isValid("OGZP0N77444Y40");
 */
function isValid(cnpj) {
    const result = cnpj.replace(/[^a-zA-Z0-9]/g, "");
    if (result.length === CNPJ_LENGTH) {
        if (isNumeric(result)) {
            return validateNumeric(result);
        }
        else if (isAlfanumeric(result)) {
            return validateAlfanumeric(result);
        }
    }
    return false;
}
/**
 * Verifica se uma string consiste apenas em caracteres numéricos (0-9).
 *
 * @param value - A string a ser verificada.
 *
 * @returns {boolean} Retorna `true` se a string for composta apenas por números, `false` caso contrário.
 *
 * @example
 * isNumeric("123456"); // true
 * isNumeric("12A345"); // false
 */
function isNumeric(value) {
    return /^[0-9]+$/.test(value);
}
/**
 * Verifica se uma string consiste apenas em caracteres alfanuméricos (A-Z, a-z, 0-9).
 *
 * @param value - A string a ser verificada.
 *
 * @returns {boolean} Retorna `true` se a string for composta apenas por caracteres alfanuméricos,
 *         `false` caso contrário.
 *
 * @example
 * isAlfanumeric("abc123"); // true
 * isAlfanumeric("abc!123"); // false
 */
function isAlfanumeric(value) {
    return /^[a-zA-Z0-9]+$/.test(value);
}
/**
 * Limpa o CNPJ removendo caracteres não numéricos, ou pontuação, dependendo do tipo de CNPJ (numérico ou alfanumérico).
 *
 * @param cnpj - O CNPJ a ser limpo.
 * @param numeric - Se `true`, remove todos os caracteres não numéricos; se `false`, remove pontuação (como pontos, barras, etc).
 *
 * @returns {string} Retorna o CNPJ limpo.
 *
 * @example
 * cleanCNPJ("25.143.815/0001-50", true); // "25143815000150"
 * cleanCNPJ("OG.ZP0.N77/444Y-42", false); // "OGZP0N77444Y42"
 */
function cleanCNPJ(cnpj, numeric) {
    return numeric ? cnpj.replace(/\D/g, "") : cnpj.replace(/[./-]/g, "");
}
/**
 * Verifica se o CNPJ tem o formato correto, com o comprimento esperado e sem sequências repetidas de dígitos.
 *
 * @param cnpj - O CNPJ a ser verificado.
 *
 * @returns {boolean} Retorna `true` se o CNPJ tiver o formato válido, `false` caso contrário.
 *
 * @example
 * isValidFormat("25143815000150"); // true
 * isValidFormat("11111111111111"); // false (sequência de números repetidos)
 */
function isValidFormat(cnpj) {
    return cnpj.length === CNPJ_LENGTH && !/^(\d)\1+$/.test(cnpj);
}
/**
 * Converte um CNPJ alfanumérico para um array de números, onde as letras são convertidas para números
 * de acordo com uma tabela de mapeamento definida.
 *
 * @param cnpj - O CNPJ alfanumérico a ser convertido (ex: "OG.ZP0.N77/444Y-42").
 *
 * @returns {number[]} Retorna um array de números representando o CNPJ, com as letras convertidas para números.
 *
 * @example
 * convertCnpj("OG.ZP0.N77/444Y-42");
 * // Retorna: [31, 32, 0, 15, 29, 42, 0, 39, 7, 7, 4, 4, 4, 31, 42]
 */
function convertCnpj(cnpj) {
    return Array.from(cnpj).map((item) => isNaN(Number(item)) ? convertLetterToNumber(item) : +item);
}
exports.default = {
    generate,
    isValid,
    format,
};
