// @dmalbuquerque/cpf-cnpj-validator/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.documentValidator = exports.cnpj = exports.cpf = void 0;
const cpf_1 = __importDefault(require("./cpf"));
exports.cpf = cpf_1.default;
const cnpj_1 = __importDefault(require("./cnpj"));
exports.cnpj = cnpj_1.default;
const documentValidator = (joi) => ({
    type: "document",
    base: joi.string().allow(null).messages({
        "string.empty": "{{#label}} não pode ser vazio",
        "any.required": "{{#label}} é obrigatório",
    }),
    messages: {
        "document.cpf": "CPF inválido",
        "document.cnpj": "CNPJ inválido",
    },
    rules: {
        cpf: {
            validate(value, helpers) {
                if (value !== null && !cpf_1.default.isValid(value))
                    return helpers.error("document.cpf");
                return value;
            },
        },
        cnpj: {
            validate(value, helpers) {
                if (value !== null && !cnpj_1.default.isValid(value))
                    return helpers.error("document.cnpj");
                return value;
            },
        },
    },
});
exports.documentValidator = documentValidator;
exports.default = exports.documentValidator;
