// @dmalbuquerque/cpf-cnpj-validator/dist/cpf/cpf.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generate = generate;
exports.isValid = isValid;
const constants_1 = require("../constants");
/**
 * Gera um número de CPF.
 *
 * Esta função gera um CPF válido. O CPF gerado pode ser retornado com ou sem a formatação.
 *
 * @param {CPFOptions} [options] - Opções para a geração do CPF.
 * @param {boolean} [options.formatted] - Se verdadeiro, retorna o CPF formatado (com pontos e traço).
 * Caso contrário, retorna o CPF sem formatação (apenas números).
 *
 * @returns {string} O CPF gerado, podendo ser formatado ou não, conforme as opções fornecidas.
 *
 * @example
 * // Retorna um CPF sem formatação, como '07208766053'
 * generate();
 *
 * @example
 * // Retorna um CPF formatado, como '072.087.660-53'
 * generate({ formatted: true });
 */
function generate(options) {
    const baseCPF = generateBaseCPF();
    const firstDigit = calculateCheckDigit(baseCPF);
    const secondDigit = calculateCheckDigit([...baseCPF, firstDigit]);
    const cpf = [...baseCPF, firstDigit, secondDigit].join("");
    return options?.formatted ? format(cpf) : cpf;
}
/**
 * Validates a CPF.
 * @param cpf The CPF to validate.
 */
function isValid(cpf) {
    const cleanedCPF = cleanCPF(cpf);
    if (!isValidFormat(cleanedCPF))
        return false;
    const digits = cleanedCPF.split("").map(Number);
    const baseCPF = digits.slice(0, 9);
    const firstDigit = calculateCheckDigit(baseCPF);
    const secondDigit = calculateCheckDigit([...baseCPF, firstDigit]);
    return firstDigit === digits[9] && secondDigit === digits[10];
}
/**
 * Cleans the CPF by removing non-numeric characters.
 * @param cpf The CPF string.
 */
function cleanCPF(cpf) {
    return cpf.replace(/\D/g, "");
}
/**
 * Checks if the CPF has a valid format and is not a sequence of repeated digits.
 * @param cpf The cleaned CPF string.
 */
function isValidFormat(cpf) {
    if (cpf.length !== constants_1.CPF_LENGTH)
        return false;
    return !/^(\d)\1+$/.test(cpf) && cpf !== "12345678909"; // Reject sequences like 111.111.111-11
}
/**
 * Generates the base CPF (first 9 digits) with random values.
 */
function generateBaseCPF() {
    return Array.from({ length: 9 }, () => Math.floor(Math.random() * 10));
}
/**
 * Calculates a CPF check digit based on the provided numbers.
 * @param numbers The array of numbers to calculate the check digit from.
 */
function calculateCheckDigit(numbers) {
    const weight = numbers.length + 1;
    const sum = numbers.reduce((acc, num, index) => acc + num * (weight - index), 0);
    const remainder = sum % 11;
    return remainder < 2 ? 0 : 11 - remainder;
}
/**
 * Formats a CPF into the pattern XXX.XXX.XXX-XX.
 * @param cpf The cleaned CPF string.
 */
function format(cpf) {
    return cpf.replace(/(\d{3})(\d{3})(\d{3})(\d{2})/, "$1.$2.$3-$4");
}
exports.default = {
    generate,
    isValid,
    format,
};
